File: src/go/fiducia_cancel_test.go
package oreslocks

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"reflect"
	"sync"
	"testing"
	"time"
)

type fiduciaRecordedCall struct {
	Path string
	Body map[string]any
}

func writeFiduciaOutput(w http.ResponseWriter, status int, output map[string]any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(map[string]any{"result": map[string]any{"output": output}})
}

func decodeFiduciaBody(t *testing.T, r *http.Request) map[string]any {
	t.Helper()
	var body map[string]any
	decoder := json.NewDecoder(r.Body)
	decoder.UseNumber()
	if err := decoder.Decode(&body); err != nil {
		t.Fatalf("decode body: %v", err)
	}
	return body
}

func requireLockKind(t *testing.T, err error, kind Kind) *Error {
	t.Helper()
	lockErr, ok := err.(*Error)
	if !ok {
		t.Fatalf("expected *Error, got %T: %v", err, err)
	}
	if lockErr.Kind != kind {
		t.Fatalf("expected %s, got %s: %v", kind, lockErr.Kind, lockErr)
	}
	return lockErr
}

func TestFiduciaTimeoutCancelsExactQueuedRequest(t *testing.T) {
	var calls []fiduciaRecordedCall
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		body := decodeFiduciaBody(t, r)
		calls = append(calls, fiduciaRecordedCall{Path: r.URL.Path, Body: body})
		switch r.URL.Path {
		case "/v1/locks/acquire":
			writeFiduciaOutput(w, http.StatusOK, map[string]any{"acquired": false})
		case "/v1/locks/cancel":
			writeFiduciaOutput(w, http.StatusOK, map[string]any{"cancelled": true, "acquired": false})
		default:
			t.Fatalf("unexpected path %s", r.URL.Path)
		}
	}))
	defer server.Close()

	lease := NewFiduciaBearer(server.URL, "test-token")
	key := LockKey("tenant/acme/jobs/rebuild")
	opts := AcquireOptions{
		TTL:           time.Minute,
		WaitTimeout:   5 * time.Millisecond,
		RetryInterval: 10 * time.Millisecond,
		Holder:        "worker-a",
		RequestID:     "attempt-42",
	}
	_, err := lease.Acquire(context.Background(), key, opts, true)
	lockErr := requireLockKind(t, err, KindTimeout)
	if !lockErr.Retryable() {
		t.Fatal("safe timeout after successful cancellation must remain retryable")
	}
	if len(calls) != 2 || calls[0].Path != "/v1/locks/acquire" || calls[1].Path != "/v1/locks/cancel" {
		t.Fatalf("unexpected calls: %#v", calls)
	}
	if calls[0].Body["request_id"] != "attempt-42" || calls[1].Body["request_id"] != "attempt-42" {
		t.Fatalf("request identity drift: %#v", calls)
	}
	if calls[0].Body["holder"] != calls[1].Body["holder"] {
		t.Fatalf("holder drift: %#v", calls)
	}
	if !reflect.DeepEqual(calls[1].Body["keys"], []any{"tenant/acme/jobs/rebuild"}) {
		t.Fatalf("cancel keys drift: %#v", calls[1].Body["keys"])
	}
}

func TestFiduciaCallerCancelReleasesRacedGrant(t *testing.T) {
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()
	var mu sync.Mutex
	var calls []fiduciaRecordedCall
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		body := decodeFiduciaBody(t, r)
		mu.Lock()
		calls = append(calls, fiduciaRecordedCall{Path: r.URL.Path, Body: body})
		mu.Unlock()
		switch r.URL.Path {
		case "/v1/locks/acquire":
			cancel()
			writeFiduciaOutput(w, http.StatusOK, map[string]any{"acquired": false})
		case "/v1/locks/cancel":
			writeFiduciaOutput(w, http.StatusOK, map[string]any{
				"cancelled": false,
				"acquired":  true,
				"grant": map[string]any{
					"holder":        "worker-a",
					"fencing_token": "41",
				},
			})
		case "/v1/locks/release":
			writeFiduciaOutput(w, http.StatusOK, map[string]any{"released": true})
		default:
			t.Fatalf("unexpected path %s", r.URL.Path)
		}
	}))
	defer server.Close()

	lease := NewFiduciaBearer(server.URL, "test-token")
	_, err := lease.Acquire(ctx, LockKey("tenant/acme/jobs/rebuild"), AcquireOptions{
		TTL:           time.Minute,
		WaitTimeout:   time.Second,
		RetryInterval: 50 * time.Millisecond,
		Holder:        "worker-a",
		RequestID:     "attempt-42",
	}, true)
	lockErr := requireLockKind(t, err, KindTransport)
	if lockErr.Retryable() {
		t.Fatal("caller cancellation must not become an automatic retry signal")
	}

	mu.Lock()
	defer mu.Unlock()
	paths := make([]string, len(calls))
	for i, call := range calls {
		paths[i] = call.Path
	}
	if !reflect.DeepEqual(paths, []string{"/v1/locks/acquire", "/v1/locks/cancel", "/v1/locks/release"}) {
		t.Fatalf("unexpected calls: %#v", calls)
	}
	if calls[1].Body["request_id"] != "attempt-42" {
		t.Fatalf("cancel request_id drift: %#v", calls[1].Body)
	}
	if calls[2].Body["holder"] != "worker-a" || calls[2].Body["fencing_token"] != json.Number("41") {
		t.Fatalf("raced grant authority drift: %#v", calls[2].Body)
	}
}

func TestFiduciaRacedGrantReleaseNoopIsSafetyFailure(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		switch r.URL.Path {
		case "/v1/locks/acquire":
			writeFiduciaOutput(w, http.StatusOK, map[string]any{"acquired": false})
		case "/v1/locks/cancel":
			writeFiduciaOutput(w, http.StatusOK, map[string]any{
				"cancelled": false,
				"acquired":  true,
				"grant": map[string]any{"holder": "worker-a", "fencing_token": "41"},
			})
		case "/v1/locks/release":
			writeFiduciaOutput(w, http.StatusOK, map[string]any{"released": false})
		default:
			t.Fatalf("unexpected path %s", r.URL.Path)
		}
	}))
	defer server.Close()

	lease := NewFiduciaBearer(server.URL, "test-token")
	_, err := lease.Acquire(context.Background(), LockKey("tenant/acme/jobs/rebuild"), AcquireOptions{
		TTL: time.Minute, WaitTimeout: time.Millisecond, RetryInterval: 2 * time.Millisecond,
		Holder: "worker-a", RequestID: "attempt-42",
	}, true)
	lockErr := requireLockKind(t, err, KindTransport)
	if lockErr.Retryable() {
		t.Fatal("unsafe raced-grant cleanup must not be retryable")
	}
}

func TestFiduciaCancelTransportFailureNeverBecomesContentionOrTimeout(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		switch r.URL.Path {
		case "/v1/locks/acquire":
			writeFiduciaOutput(w, http.StatusOK, map[string]any{"acquired": false})
		case "/v1/locks/cancel":
			writeFiduciaOutput(w, http.StatusServiceUnavailable, map[string]any{"error": "unavailable"})
		default:
			t.Fatalf("unexpected path %s", r.URL.Path)
		}
	}))
	defer server.Close()

	lease := NewFiduciaBearer(server.URL, "test-token")
	_, err := lease.Acquire(context.Background(), LockKey("tenant/acme/jobs/rebuild"), AcquireOptions{
		TTL: time.Minute, WaitTimeout: time.Millisecond, RetryInterval: 2 * time.Millisecond,
		Holder: "worker-a", RequestID: "attempt-42",
	}, true)
	lockErr := requireLockKind(t, err, KindTransport)
	if lockErr.Retryable() {
		t.Fatal("ambiguous cancellation must fail closed as non-retryable transport safety failure")
	}
}

File: src/ts/test/fiducia-cancel.test.mjs
import assert from "node:assert/strict";
import test from "node:test";

import { FiduciaLease, LockError, lockKey } from "../dist/index.js";

function response(output, status = 200) {
  return {
    status,
    async text() {
      return JSON.stringify({ result: { output } });
    },
  };
}

const key = lockKey("tenant/acme/jobs/rebuild");
const baseOpts = {
  ttlMs: 60_000,
  waitTimeoutMs: 5,
  retryIntervalMs: 10,
  holder: "worker-a",
  requestId: "attempt-42",
};

test("Fiducia timeout cancels the exact queued request before returning timeout", async () => {
  const calls = [];
  const lease = new FiduciaLease({
    baseUrl: "https://fiducia.example.test",
    fetch: async (url, init) => {
      const path = new URL(url).pathname;
      calls.push({ path, body: JSON.parse(init.body) });
      if (path === "/v1/locks/acquire") return response({ acquired: false });
      if (path === "/v1/locks/cancel") return response({ cancelled: true, acquired: false });
      throw new Error(`unexpected path ${path}`);
    },
  });

  await assert.rejects(
    () => lease.acquire(key, baseOpts, true),
    (error) => error instanceof LockError && error.kind === "timeout" && error.retryable,
  );

  assert.deepEqual(calls.map((call) => call.path), ["/v1/locks/acquire", "/v1/locks/cancel"]);
  assert.equal(calls[0].body.request_id, "attempt-42");
  assert.equal(calls[1].body.request_id, "attempt-42");
  assert.equal(calls[0].body.holder, calls[1].body.holder);
  assert.deepEqual(calls[1].body.keys, [String(key)]);
});

test("Fiducia caller cancellation releases a grant won by the cancel race", async () => {
  const calls = [];
  const controller = new AbortController();
  const lease = new FiduciaLease({
    baseUrl: "https://fiducia.example.test",
    fetch: async (url, init) => {
      const path = new URL(url).pathname;
      calls.push({ path, body: JSON.parse(init.body) });
      if (path === "/v1/locks/acquire") {
        controller.abort(new Error("caller stopped waiting"));
        return response({ acquired: false });
      }
      if (path === "/v1/locks/cancel") {
        return response({
          cancelled: false,
          acquired: true,
          grant: { holder: "worker-a", fencing_token: "41" },
        });
      }
      if (path === "/v1/locks/release") return response({ released: true });
      throw new Error(`unexpected path ${path}`);
    },
  });

  await assert.rejects(
    () => lease.acquire(key, { ...baseOpts, signal: controller.signal }, true),
    (error) => error instanceof LockError && error.kind === "transport" && !error.retryable,
  );

  assert.deepEqual(calls.map((call) => call.path), [
    "/v1/locks/acquire",
    "/v1/locks/cancel",
    "/v1/locks/release",
  ]);
  assert.equal(calls[0].body.request_id, "attempt-42");
  assert.equal(calls[1].body.request_id, "attempt-42");
  assert.equal(calls[2].body.holder, "worker-a");
  assert.equal(calls[2].body.fencing_token, 41);
});

test("Fiducia raced-grant release no-op becomes a safety failure", async () => {
  const lease = new FiduciaLease({
    baseUrl: "https://fiducia.example.test",
    fetch: async (url) => {
      const path = new URL(url).pathname;
      if (path === "/v1/locks/acquire") return response({ acquired: false });
      if (path === "/v1/locks/cancel") {
        return response({
          cancelled: false,
          acquired: true,
          grant: { holder: "worker-a", fencing_token: "41" },
        });
      }
      if (path === "/v1/locks/release") return response({ released: false });
      throw new Error(`unexpected path ${path}`);
    },
  });

  await assert.rejects(
    () => lease.acquire(key, baseOpts, true),
    (error) =>
      error instanceof LockError &&
      error.kind === "transport" &&
      /raced grant release was a no-op/.test(error.message),
  );
});

test("Fiducia cancel transport ambiguity never downgrades to timeout or contention", async () => {
  const lease = new FiduciaLease({
    baseUrl: "https://fiducia.example.test",
    fetch: async (url) => {
      const path = new URL(url).pathname;
      if (path === "/v1/locks/acquire") return response({ acquired: false });
      if (path === "/v1/locks/cancel") return response({ error: "unavailable" }, 503);
      throw new Error(`unexpected path ${path}`);
    },
  });

  await assert.rejects(
    () => lease.acquire(key, baseOpts, true),
    (error) => error instanceof LockError && error.kind === "transport" && !error.retryable,
  );
});
